### PR TITLE
Remove dependency python dateutil

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Remove dependency python dateutil.

### Changed

- Remove dependency python dateutil
- Refactored `ls CLI`
- Refactored server: `post_query_analyses` and `find_analysis`

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b remove_python-dateutil -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
### CLI ✅ 
```
henrik.stranneheim@hasta:~] [S_base] 9s $ trailblazer ls --help
Usage: trailblazer ls [OPTIONS]

  Display recent logs for the latest analyses.

Options:
  -s, --status [canceled|completed|completing|error|failed|pending|qc|running]
                                  Find analysis with specified status
  -b, --before TEXT               Find analyses started before date
  -c, --comment TEXT              Find analysis with comment
  --help                          Show this message and exit.

271995 | cuddlyhen 2023-08-09 [WTS/FAILED]
271994 | justhusky 2023-08-08 [WGS/CANCELED]
271975 | ampleray 2023-07-25 [WTS/FAILED]
271971 | acetuna 2023-07-24 [TGS/COMPLETED] - None
271970 | uphippo 2023-07-24 [TGS/FAILED]
271969 | equalbug 2023-07-24 [TGS/FAILED]
271968 | unitedbeagle 2023-07-24 [TGS/COMPLETED] - None
271967 | setamoeba 2023-07-24 [TGS/COMPLETED] - None
271966 | fleetjay 2023-07-24 [WGS/FAILED]
271965 | civilsole 2023-07-24 [WGS/FAILED]
271957 | uphippo_rsync 2023-07-17 [OTHER/FAILED]
271955 | setamoeba_rsync 2023-07-17 [OTHER/FAILED]
271907 | equalbug_rsync 2023-07-07 [OTHER/FAILED]
271906 | equalbug_rsync 2023-07-07 [OTHER/FAILED]
271905 | equalbug_rsync 2023-07-07 [OTHER/FAILED]
271904 | H27CLDRX3 2023-06-30 [OTHER/COMPLETED] - None
271903 | intentcorgi 2023-06-27 [WGS/COMPLETED] - None
271902 | sharpwhale 2023-06-26 [WGS/COMPLETED] - None
271901 | safeguinea 2023-06-26 [WGS/COMPLETED] - None
271900 | onemite 2023-06-26 [WGS/COMPLETED] - None
271899 | brightcaiman 2023-06-26 [WGS/COMPLETED] - None
271897 | tenderoriole 2023-06-22 [TGS/COMPLETED] - None
271896 | strongbison 2023-06-22 [WES/COMPLETED] - None
271894 | sharppigeon 2023-06-22 [WGS/COMPLETED] - None
271892 | proudcougar 2023-06-22 [WGS/COMPLETED] - None
271890 | nextjackal 2023-06-22 [WGS/COMPLETED] - None
271889 | modernmule 2023-06-22 [WGS/COMPLETED] - None
271888 | intentmayfly 2023-06-22 [WGS/COMPLETED] - None
271887 | kindcaiman 2023-06-22 [WGS/COMPLETED] - None
271885 | inferret 2023-06-22 [WGS/COMPLETED] - None

[henrik.stranneheim@hasta:~] [S_base] 9s $ trailblazer ls --before "2023-07-06"
271904 | H27CLDRX3 2023-06-30 [OTHER/COMPLETED] - None
271903 | intentcorgi 2023-06-27 [WGS/COMPLETED] - None
271902 | sharpwhale 2023-06-26 [WGS/COMPLETED] - None
271901 | safeguinea 2023-06-26 [WGS/COMPLETED] - None
271900 | onemite 2023-06-26 [WGS/COMPLETED] - None
271899 | brightcaiman 2023-06-26 [WGS/COMPLETED] - None
271897 | tenderoriole 2023-06-22 [TGS/COMPLETED] - None
271896 | strongbison 2023-06-22 [WES/COMPLETED] - None
271894 | sharppigeon 2023-06-22 [WGS/COMPLETED] - None
271892 | proudcougar 2023-06-22 [WGS/COMPLETED] - None
271890 | nextjackal 2023-06-22 [WGS/COMPLETED] - None
271889 | modernmule 2023-06-22 [WGS/COMPLETED] - None
271888 | intentmayfly 2023-06-22 [WGS/COMPLETED] - None
271887 | kindcaiman 2023-06-22 [WGS/COMPLETED] - None
271885 | inferret 2023-06-22 [WGS/COMPLETED] - None
271884 | helpedfilly 2023-06-22 [WES/COMPLETED] - None
271883 | gladthrush 2023-06-22 [WES/COMPLETED] - None
271882 | cleanshrimp 2023-06-22 [WES/COMPLETED] - None
271880 | politeglider 2023-06-21 [WTS/COMPLETED] - None
271877 | 2257WYLT3 2023-06-14 [OTHER/FAILED]
271876 | setamoeba_rsync 2023-06-07 [OTHER/FAILED]
271875 | equalbug_rsync 2023-06-07 [OTHER/FAILED]
271874 | uphippo_rsync 2023-06-07 [OTHER/FAILED]
271870 | crackyak 2023-06-03 [WGS/FAILED]
271865 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271864 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271863 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271862 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271853 | casualweasel 2023-05-25 [TGS/FAILED]
271852 | eagerox 2023-05-25 [TGS/FAILED]
[henrik.stranneheim@hasta:~] [S_base] $ trailblazer ls --before "2023-07-06" --status failed
271877 | 2257WYLT3 2023-06-14 [OTHER/FAILED]
271876 | setamoeba_rsync 2023-06-07 [OTHER/FAILED]
271875 | equalbug_rsync 2023-06-07 [OTHER/FAILED]
271874 | uphippo_rsync 2023-06-07 [OTHER/FAILED]
271870 | crackyak 2023-06-03 [WGS/FAILED]
271865 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271864 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271863 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271862 | politeglider_rsync 2023-06-01 [OTHER/FAILED]
271853 | casualweasel 2023-05-25 [TGS/FAILED]
271852 | eagerox 2023-05-25 [TGS/FAILED]
271851 | steadyhound 2023-05-23 [TGS/FAILED]
271850 | civilsole_rsync 2023-05-22 [OTHER/FAILED]
271849 | fleetjay_rsync 2023-05-22 [OTHER/FAILED]
271848 | fleetjay_rsync 2023-05-22 [OTHER/FAILED]
271846 | equalbug_rsync 2023-05-22 [OTHER/FAILED]
271845 | uphippo_rsync 2023-05-22 [OTHER/FAILED]
271841 | 231889 2023-05-22 [OTHER/FAILED]
271840 | newhound_rsync 2023-05-22 [OTHER/FAILED]
271835 | uphippo_rsync 2023-05-22 [OTHER/FAILED]
271833 | uphippo_rsync 2023-05-22 [OTHER/FAILED]
271832 | brightsheep 2023-05-17 [OTHER/FAILED]
271825 | upwardstork 2023-05-11 [WGS/FAILED]
271821 | suitedgrub 2023-05-11 [WGS/FAILED]
271818 | notedshark 2023-05-11 [WGS/FAILED]
271816 | largeturtle 2023-05-11 [WGS/FAILED]
271808 | setamoeba_rsync 2023-05-10 [OTHER/FAILED]
271807 | cuddlyhen_partial 2023-05-10 [OTHER/FAILED]
271774 | nearbyjoey 2023-04-26 [WTS/FAILED]
```

### API ✅ 
```
56f4e0d892fe  docker.io/clinicalgenomics/trailblazer-stage:remove_python-dateutil            /bin/sh -c gunico...  26 seconds ago  Up 26 seconds ago  0.0.0.0:7076->8000/tcp  trailblazer-remove_python-dateutil
```
Screenshot 2023-08-09 at 11 08 10

File: trailblazer/server/api.py [hunk 1]
```diff
@@ -166,21 +165,25 @@ def delete(analysis_id):
...
+        family=post_request.get("family"),
+        is_visible=post_request.get("visible"),
+        query=post_request.get("query"),
+        status=post_request.get("status"),
+        temp=post_request.get("temp"),
     )
-    data = [stringify_timestamps(analysis_obj.to_dict()) for analysis_obj in query_analyses]
-    return jsonify(*data), 200
+    raw_analyses: List[Dict[str, str]] = [
+        stringify_timestamps(analysis_obj.to_dict()) for analysis_obj in query_analyses
```
Comment (fevac) [inline trailblazer/server/api.py:184]:
```suggestion
        stringify_timestamps(analysis.to_dict()) for analysis in query_analyses
```

**Review by fevac — CHANGES_REQUESTED** [review]
Nice! 🌟 I just had minor comments. Is the point to have less dependencies?

File: trailblazer/server/api.py [hunk 2]
```diff
@@ -166,21 +165,25 @@ def delete(analysis_id):
 
 @blueprint.route("/query-analyses", methods=["POST"])
 def post_query_analyses():
-    """Return list of analyses matching the query terms"""
-    content = request.json
+    """Return list of analyses matching the query terms."""
+    post_request: Response.json = request.json
     query_analyses = store.analyses(
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Done

File: trailblazer/server/api.py [hunk 3]
```diff
@@ -197,15 +200,15 @@ def post_get_latest_analysis():
...
     """Find analysis using case_id, date, and status"""
-    content = request.json
-    analysis_obj = store.get_analysis(
-        case_id=content.get("case_id"),
-        started_at=parse_datestr(content.get("started_at")),
-        status=content.get("status"),
+    post_request: Response.json = request.json
+    analysis: Analysis = store.get_analysis(
+        case_id=post_request.get("case_id"),
+        started_at=datetime.strptime(post_request.get("started_at"), "%Y-%m-%d").date(),
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Done

File: trailblazer/server/api.py [hunk 4]
```diff
@@ -197,15 +200,15 @@ def post_get_latest_analysis():
 @blueprint.route("/find-analysis", methods=["POST"])
 def post_find_analysis():
     """Find analysis using case_id, date, and status"""
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Fixed

File: trailblazer/constants.py [hunk 5]
```diff
@@ -95,6 +95,15 @@ def ongoing_statuses(cls) -> Tuple:
         return cls.PENDING.value, cls.RUNNING.value, cls.COMPLETING.value, cls.ERROR.value
 
 
+class TrailblazerStatusColor(str, Enum):
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/constants.py:99]: Let's add them when needed

File: trailblazer/cli/utils/ls_helper.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+from typing import Dict
+
+from trailblazer.constants import TrailblazerStatus, TrailblazerStatusColor
+from trailblazer.store.models import Analysis
+
+
+def _get_ls_analysis_message(analysis: Analysis) -> tuple:
+    """Return the log message for the analysis."""
+    message = (
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/cli/utils/ls_helper.py]: Done

File: trailblazer/cli/utils/ls_helper.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from typing import Dict
+
+from trailblazer.constants import TrailblazerStatus, TrailblazerStatusColor
+from trailblazer.store.models import Analysis
+
+
+def _get_ls_analysis_message(analysis: Analysis) -> tuple:
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/cli/utils/ls_helper.py]: Done

File: trailblazer/constants.py [hunk 8]
```diff
@@ -95,6 +95,15 @@ def ongoing_statuses(cls) -> Tuple:
...
+    """Trailblazer status colors."""
+
+    COMPLETED: str = "green"
+    FAILED: str = "red"
+    PENDING: str = "yellow"
+    RUNNING: str = "blue"
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/constants.py:106]: Added

File: trailblazer/server/api.py (hunk 1 above)
Comment (henrikstranneheim) [reply trailblazer/server/api.py:185]: There is an outer scope warning for this refactoring. Will do it when the server api is refactored

File: trailblazer/server/api.py [hunk 9]
```diff
@@ -197,15 +200,15 @@ def post_get_latest_analysis():
...
+        case_id=post_request.get("case_id"),
+        started_at=datetime.strptime(post_request.get("started_at"), "%Y-%m-%d").date(),
+        status=post_request.get("status"),
     )
-    if analysis_obj:
-        data = stringify_timestamps(analysis_obj.to_dict())
-        return jsonify(**data), 200
+    if analysis:
+        raw_analysis: Dict[str, str] = stringify_timestamps(analysis.to_dict())
+        return jsonify(**raw_analysis), 200
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Done

**Comment by henrikstranneheim** [comment]
> I just had minor comments. Is the point to have less dependencies?

Yes, if they do not add significant value. Non-core deps couples the software to another codebase making it more vulnerable to changes.

**Comment by henrikstranneheim** [comment]
@fevac Ready for another review

**Review by fevac — APPROVED** [review]
🚀

**Comment by henrikstranneheim** [comment]
Reran stage test which passed for both CLI and API

**Comment by henrikstranneheim** [comment]
Deployed to production:
```
Log deploy... done.
trailblazer, version 15.2.0
θ69° [henrik.stranneheim@hasta:~] [P_base] 1m18s $
```

**Pull request merged**